flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceSplitReader.java

```diff
@@ -80,6 +87,18 @@ public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
     }
 
     if (currentReader.hasNext()) {
+      // Wait until the reader is blocked. Wake every second in case this missed a signal
+      while (pausedSplits.contains(currentSplitId)) {
```

Comment:
do we need some sort of timeout here ?

Reply:
If the unpause is not called then we should wait. Maybe logging should be good. Info on first loop, debug/trace on consecutive loops.

Reply:
I am not sure this implements the contract correctly.

```
    /**
     * Fetch elements into the blocking queue for the given splits. The fetch call could be blocking
     * but it should get unblocked when {@link #wakeUp()} is invoked. In that case, the
     * implementation may either decide to return without throwing an exception, or it can just
     * throw an interrupted exception. In either case, this method should be reentrant, meaning that
     * the next fetch call should just resume from where the last fetch call was waken up or
     * interrupted. It is up to the implementer to either read all the records of the split or to
     * stop reading them at some point (for example when a given threshold is exceeded). In that
     * later case, when fetch is called again, the reading should restart at the record where it
     * left off based on the {@code SplitState}.
     *
     * @return the Ids of the finished splits.
     * @throws IOException when encountered IO errors, such as deserialization failures.
     */
    RecordsWithSplitIds<E> fetch() throws IOException;
```